File: engines/swift/Sources/DivorceCore/Divide.swift
public func divide(
    _ caseDoc: MaritalPropertyCase,
    scenarioId: String,
    options: ComputeOptions = ComputeOptions()
) throws -> DivisionResult {
    guard let scenario = caseDoc.scenarios.first(where: { $0.id == scenarioId }) else {
        throw EngineError.scenarioNotFound(scenarioId)
    }
    guard caseDoc.regime == "ru_community_property" else {
        throw EngineError.regimeNotSupported
    }

    let warnings = WarningSink()
    let plaintiff = scenario.claim?.plaintiff ?? caseDoc.facts.parties.plaintiff
    let childrenCount = caseDoc.facts.children.count
    let cutoff: String? = scenario.massCutoff == .dissolution
        ? caseDoc.facts.dissolvedAt
        : caseDoc.facts.separatedAt
    let asOf = options.asOf ?? scenario.claim?.filedAt

    if scenario.wifeShare != 0.5 && (scenario.deviationGrounds ?? []).isEmpty {
        warnings.add("scenario.deviation_without_grounds")
    }
    if caseDoc.facts.agreement?.exists == true && caseDoc.facts.agreement?.notarized == false {
        warnings.add("agreement.not_notarized")
    }

    var assetsCommon = 0.0
    var receivedAssets = SideAmounts()
    var unallocated = 0.0
    var claimedByPlaintiff = 0.0

    for item in scenario.assetItems {
        guard item.include else { continue }
        guard let asset = caseDoc.assets.first(where: { $0.id == item.assetId }) else {
            warnings.add("asset.missing")
            continue
        }
        if excludedByAgreement(caseDoc, assetId: asset.id) { continue }

        if let cutoff, let acquired = asset.acquiredAt, acquired > cutoff {
            warnings.add("asset.acquired_after_separation")
            if item.acceptedPosition == nil { continue }
        }

        collectAssetWarnings(asset, warnings)

        let position = resolvePosition(
            asset: asset,
            item: item,
            scenario: scenario,
            plaintiff: plaintiff,
            warnings: warnings
        )
        guard inMass(position) else { continue }

        let value = assetValueRubles(asset: asset, item: item, scenario: scenario, warnings: warnings, asOf: asOf)
        let personal = item.personalShareOverride
            ?? chosenPersonalShare(asset: asset, position: position, item: item)
            ?? fundingPersonalShare(asset.funding)
        let kids = childrenShare(asset: asset, childrenCount: childrenCount)
        if n(asset.funding?.maternityCapital) > 0 {
            warnings.add("asset.maternity_capital_children_share")
        }
        let commonPart = value * max(0, 1 - personal - kids)
        assetsCommon += commonPart

        let addBack = item.addBack == true && asset.disposed != nil
        if let plaintiff, item.assignedTo == transfer(plaintiff), !addBack {
            claimedByPlaintiff += commonPart
        }
        if addBack {
            if let disposer = asset.disposed?.by {
                receivedAssets[disposer] += commonPart
            } else {
                unallocated += commonPart
            }
        } else if let assigned = item.assignedTo, assigned == .wife || assigned == .husband {
            receivedAssets[side(assigned)] += commonPart
        } else if item.assignedTo == .jointShares, let ratio = item.shareRatio {
            receivedAssets.wife += commonPart * ratio.wife
            receivedAssets.husband += commonPart * ratio.husband
        } else {
            unallocated += commonPart
        }
    }

    var liabilitiesCommon = 0.0
    var receivedLiabilities = SideAmounts()

    for item in scenario.liabilityItems {
        guard item.include else { continue }
        guard let liability = caseDoc.liabilities.first(where: { $0.id == item.liabilityId }) else {
            warnings.add("liability.missing")
            continue
        }
        if liability.familyPurpose.status != .proven {
            warnings.add("liability.family_purpose_not_proven")
        }
        guard item.treatAsCommon else { continue }

        guard let amount = toRubles(
            amount: liability.outstanding.amount,
            currency: liability.currency,
            fxRate: liability.fxRate
        ) else {
            warnings.add("valuation.fx_missing")
            continue
        }
        liabilitiesCommon += amount
        if liability.kind == "mortgage" {
            warnings.add("liability.bank_consent_required")
        }
        switch item.allocation {
        case .wife:
            receivedLiabilities.wife += amount
        case .husband:
            receivedLiabilities.husband += amount
        case .proportional:
            receivedLiabilities.wife += amount * scenario.wifeShare
            receivedLiabilities.husband += amount * (1 - scenario.wifeShare)
        }
    }

    let net = assetsCommon - liabilitiesCommon
    let ideal = SideAmounts(wife: net * scenario.wifeShare, husband: net * (1 - scenario.wifeShare))
    let receivedNet = SideAmounts(
        wife: receivedAssets.wife - receivedLiabilities.wife,
        husband: receivedAssets.husband - receivedLiabilities.husband
    )
    let compensation = compensationOf(receivedNet: receivedNet, ideal: ideal)
    if unallocated > 0 { warnings.add("division.unallocated_present") }

    let claimPrice = claimPriceOf(
        scenario: scenario,
        plaintiff: plaintiff,
        claimedByPlaintiff: claimedByPlaintiff,
        compensation: compensation
    )
    var fee: Double?
    var court: Jurisdiction?
    if let claimPrice {
        let feeResult = try courtFee(CourtFeeInput(
            claimPrice: claimPrice,
            filedAt: scenario.claim?.filedAt,
            plaintiffType: scenario.claim?.plaintiffType ?? .individual,
            divorceClaimed: scenario.claim?.divorceClaimed ?? false,
            priorRightEstablished: scenario.claim?.priorRightEstablished ?? false,
            injunctionRequested: scenario.claim?.injunctionRequested ?? false
        ))
        fee = feeResult.courtFee
        court = jurisdiction(claimPrice)
        for code in feeResult.warnings { warnings.add(code) }
    }

    return DivisionResult(
        mass: MassTotals(assetsCommon: assetsCommon, liabilitiesCommon: liabilitiesCommon, net: net),
        ideal: ideal,
        received: ReceivedTotals(assets: receivedAssets, liabilities: receivedLiabilities, net: receivedNet),
        unallocated: unallocated,
        compensation: compensation,
        claimPrice: claimPrice,
        courtFee: fee,
        jurisdiction: court,
        warnings: warnings.codes
    )
}

public func claimPrice(
    _ caseDoc: MaritalPropertyCase,
    scenarioId: String,
    options: ComputeOptions = ComputeOptions()
) throws -> Double? {
    try divide(caseDoc, scenarioId: scenarioId, options: options).claimPrice
}

public func warningsFor(
    _ caseDoc: MaritalPropertyCase,
    scenarioId: String,
    options: ComputeOptions = ComputeOptions()
) throws -> [String] {
    try divide(caseDoc, scenarioId: scenarioId, options: options).warnings
}

private func transfer(_ side: Side) -> TransferTarget {
    side == .wife ? .wife : .husband
}

private func side(_ target: TransferTarget) -> Side {
    target == .wife ? .wife : .husband
}

private func excludedByAgreement(_ caseDoc: MaritalPropertyCase, assetId: String) -> Bool {
    guard let agreement = caseDoc.facts.agreement, agreement.exists else { return false }
    if agreement.notarized == false { return false }
    return agreement.assetIds?.contains(assetId) ?? false
}

private func chosenPersonalShare(asset: Asset, position: Position, item: ScenarioAssetItem) -> Double? {
    if item.acceptedPosition == .partlyPersonal || position == .partlyPersonal {
        return asset.positions.wife.claimedPersonalShare ?? asset.positions.husband.claimedPersonalShare
    }
    return nil
}

private func compensationOf(receivedNet: SideAmounts, ideal: SideAmounts) -> Compensation {
    let diff = receivedNet.wife - ideal.wife
    if abs(diff) < 0.005 { return Compensation(from: nil, to: nil, amount: 0) }
    if diff > 0 { return Compensation(from: .wife, to: .husband, amount: diff) }
    return Compensation(from: .husband, to: .wife, amount: -diff)
}

private func claimPriceOf(
    scenario: Scenario,
    plaintiff: Side?,
    claimedByPlaintiff: Double,
    compensation: Compensation
) -> Double? {
    if let override = scenario.claim?.claimPriceOverride { return override }
    guard let plaintiff else { return nil }
    var price = claimedByPlaintiff
    if compensation.to == plaintiff { price += compensation.amount }
    return price
}

private func collectAssetWarnings(_ asset: Asset, _ warnings: WarningSink) {
    let wife = asset.positions.wife
    let husband = asset.positions.husband
    if wife.position == .disputed && husband.position == .disputed {
        warnings.add("asset.no_position_other_side")
    }
    func isPersonal(_ side: SidePosition) -> Bool {
        side.position == .personalWife || side.position == .personalHusband
    }
    if (isPersonal(wife) && (wife.ground == nil || wife.ground?.isEmpty == true))
        || (isPersonal(husband) && (husband.ground == nil || husband.ground?.isEmpty == true))
    {
        warnings.add("asset.personal_without_ground")
    }
    if asset.disposed?.spouseConsent == false {
        warnings.add("asset.disposed_without_evidence")
    }
    if (wife.position == .partlyPersonal || husband.position == .partlyPersonal)
        && n(asset.funding?.personalWife) == 0
        && n(asset.funding?.personalHusband) == 0
    {
        warnings.add("asset.partly_personal_needs_tracing")
    }
    if asset.kind == "business_share" {
        warnings.add("business.charter_consent_required")
    }
    if asset.kind == "crypto" {
        warnings.add("asset.crypto_evidence")
    }
}

File: engines/ts/src/divide.ts
import { courtFee } from "./courtFee.ts";
import { jurisdiction } from "./jurisdiction.ts";
import { n } from "./money.ts";
import type {
  Asset,
  Compensation,
  ComputeOptions,
  DivisionResult,
  MaritalPropertyCase,
  Scenario,
  Side,
  SideAmounts,
} from "./types.ts";
import {
  assetValueRubles,
  childrenShare,
  fundingPersonalShare,
  inMass,
  resolvePosition,
  toRubles,
} from "./value.ts";

function emptySides(): SideAmounts {
  return { wife: 0, husband: 0 };
}

function addWarning(warnings: string[], code: string): void {
  if (!warnings.includes(code)) warnings.push(code);
}

function plaintiffOf(caseDoc: MaritalPropertyCase, scenario: Scenario): Side | null {
  return scenario.claim?.plaintiff ?? caseDoc.facts.parties.plaintiff ?? null;
}

function cutoffDate(caseDoc: MaritalPropertyCase, scenario: Scenario): string | null {
  return scenario.mass_cutoff === "dissolution"
    ? (caseDoc.facts.dissolved_at ?? null)
    : (caseDoc.facts.separated_at ?? null);
}

function excludedByAgreement(caseDoc: MaritalPropertyCase, assetId: string): boolean {
  const agreement = caseDoc.facts.agreement;
  if (!agreement?.exists) return false;
  if (agreement.notarized === false) {
    return false;
  }
  return agreement.asset_ids?.includes(assetId) ?? false;
}

export function divide(
  caseDoc: MaritalPropertyCase,
  scenarioId: string,
  options: ComputeOptions = {},
): DivisionResult {
  const scenario = caseDoc.scenarios.find((s) => s.id === scenarioId);
  if (!scenario) throw new Error(`scenario_not_found:${scenarioId}`);
  if (caseDoc.regime !== "ru_community_property") {
    throw new Error("regime_not_supported");
  }

  const warnings: string[] = [];
  const plaintiff = plaintiffOf(caseDoc, scenario);
  const childrenCount = caseDoc.facts.children.length;
  const cutoff = cutoffDate(caseDoc, scenario);
  const asOf = options.asOf ?? scenario.claim?.filed_at ?? null;

  if (scenario.wife_share !== 0.5 && (!scenario.deviation_grounds || scenario.deviation_grounds.length === 0)) {
    addWarning(warnings, "scenario.deviation_without_grounds");
  }
  if (caseDoc.facts.agreement?.exists && caseDoc.facts.agreement.notarized === false) {
    addWarning(warnings, "agreement.not_notarized");
  }

  let assetsCommon = 0;
  const receivedAssets = emptySides();
  let unallocated = 0;
  let claimedByPlaintiff = 0;

  for (const item of scenario.asset_items) {
    if (!item.include) continue;
    const asset = caseDoc.assets.find((a) => a.id === item.asset_id);
    if (!asset) {
      addWarning(warnings, "asset.missing");
      continue;
    }

    if (excludedByAgreement(caseDoc, asset.id)) continue;

    if (cutoff && asset.acquired_at && asset.acquired_at > cutoff) {
      addWarning(warnings, "asset.acquired_after_separation");
      if (!item.accepted_position) continue;
    }

    collectAssetWarnings(asset, warnings);

    const position = resolvePosition(asset, item, scenario, plaintiff, warnings);
    if (!inMass(position)) continue;

    const value = assetValueRubles(asset, item, scenario, warnings, { asOf });
    const personal =
      item.personal_share_override ??
      chosenPersonalShare(asset, position, item) ??
      fundingPersonalShare(asset.funding);
    const kids = childrenShare(asset, childrenCount);
    if (n(asset.funding?.maternity_capital) > 0) {
      addWarning(warnings, "asset.maternity_capital_children_share");
    }
    const commonPart = value * Math.max(0, 1 - personal - kids);
    assetsCommon += commonPart;

    const addBack = Boolean(item.add_back && asset.disposed);
    if (plaintiff && item.assigned_to === plaintiff && !addBack) {
      claimedByPlaintiff += commonPart;
    }
    if (addBack) {
      const disposer = asset.disposed?.by;
      if (disposer) receivedAssets[disposer] += commonPart;
      else unallocated += commonPart;
    } else if (item.assigned_to === "wife" || item.assigned_to === "husband") {
      receivedAssets[item.assigned_to] += commonPart;
    } else if (item.assigned_to === "joint_shares" && item.share_ratio) {
      receivedAssets.wife += commonPart * item.share_ratio.wife;
      receivedAssets.husband += commonPart * item.share_ratio.husband;
    } else {
      unallocated += commonPart;
    }
  }

  let liabilitiesCommon = 0;
  const receivedLiabilities = emptySides();

  for (const item of scenario.liability_items) {
    if (!item.include) continue;
    const liability = caseDoc.liabilities.find((l) => l.id === item.liability_id);
    if (!liability) {
      addWarning(warnings, "liability.missing");
      continue;
    }
    if (liability.family_purpose.status !== "proven") {
      addWarning(warnings, "liability.family_purpose_not_proven");
    }
    if (!item.treat_as_common) continue;

    const amount = toRubles(liability.outstanding.amount, liability.currency, liability.fx_rate);
    if (amount == null) {
      addWarning(warnings, "valuation.fx_missing");
      continue;
    }
    liabilitiesCommon += amount;
    if (liability.kind === "mortgage") {
      addWarning(warnings, "liability.bank_consent_required");
    }

    if (item.allocation === "wife") receivedLiabilities.wife += amount;
    else if (item.allocation === "husband") receivedLiabilities.husband += amount;
    else {
      receivedLiabilities.wife += amount * scenario.wife_share;
      receivedLiabilities.husband += amount * (1 - scenario.wife_share);
    }
  }

  const net = assetsCommon - liabilitiesCommon;
  const ideal: SideAmounts = {
    wife: net * scenario.wife_share,
    husband: net * (1 - scenario.wife_share),
  };
  const receivedNet: SideAmounts = {
    wife: receivedAssets.wife - receivedLiabilities.wife,
    husband: receivedAssets.husband - receivedLiabilities.husband,
  };
  const compensation = compensationOf(receivedNet, ideal);
  if (unallocated > 0) addWarning(warnings, "division.unallocated_present");

  const claimPrice = claimPriceOf(scenario, plaintiff, claimedByPlaintiff, compensation);
  let fee: number | null = null;
  let court: DivisionResult["jurisdiction"] = null;
  if (claimPrice != null) {
    const feeResult = courtFee({
      claim_price: claimPrice,
      filed_at: scenario.claim?.filed_at ?? undefined,
      plaintiff_type: scenario.claim?.plaintiff_type ?? "individual",
      divorce_claimed: scenario.claim?.divorce_claimed ?? false,
      prior_right_established: scenario.claim?.prior_right_established ?? false,
      injunction_requested: scenario.claim?.injunction_requested ?? false,
    });
    fee = feeResult.court_fee;
    court = jurisdiction(claimPrice);
    for (const w of feeResult.warnings) addWarning(warnings, w);
  }

  return {
    mass: { assets_common: assetsCommon, liabilities_common: liabilitiesCommon, net },
    ideal,
    received: { assets: receivedAssets, liabilities: receivedLiabilities, net: receivedNet },
    unallocated,
    compensation,
    claim_price: claimPrice,
    court_fee: fee,
    jurisdiction: court,
    warnings,
  };
}

function chosenPersonalShare(
  asset: Asset,
  position: ReturnType<typeof resolvePosition>,
  item: { accepted_position?: string | null },
): number | null {
  if (item.accepted_position === "partly_personal" || position === "partly_personal") {
    return (
      asset.positions.wife.claimed_personal_share ??
      asset.positions.husband.claimed_personal_share ??
      null
    );
  }
  return null;
}

function compensationOf(receivedNet: SideAmounts, ideal: SideAmounts): Compensation {
  const diff = receivedNet.wife - ideal.wife;
  if (Math.abs(diff) < 0.005) return { from: null, to: null, amount: 0 };
  if (diff > 0) return { from: "wife", to: "husband", amount: diff };
  return { from: "husband", to: "wife", amount: -diff };
}

function claimPriceOf(
  scenario: Scenario,
  plaintiff: Side | null,
  claimedByPlaintiff: number,
  compensation: Compensation,
): number | null {
  if (scenario.claim?.claim_price_override != null) return scenario.claim.claim_price_override;
  if (!plaintiff) return null;
  let price = claimedByPlaintiff;
  if (compensation.to === plaintiff) price += compensation.amount;
  return price;
}

function collectAssetWarnings(asset: Asset, warnings: string[]): void {
  const w = asset.positions.wife;
  const h = asset.positions.husband;
  if (w.position === "disputed" && h.position === "disputed") {
    addWarning(warnings, "asset.no_position_other_side");
  }
  const personal = (side: typeof w) =>
    side.position === "personal_wife" || side.position === "personal_husband";
  if ((personal(w) && !w.ground) || (personal(h) && !h.ground)) {
    addWarning(warnings, "asset.personal_without_ground");
  }
  if (asset.disposed && asset.disposed.spouse_consent === false) {
    addWarning(warnings, "asset.disposed_without_evidence");
  }
  if (
    (w.position === "partly_personal" || h.position === "partly_personal") &&
    !asset.funding?.personal_wife &&
    !asset.funding?.personal_husband
  ) {
    addWarning(warnings, "asset.partly_personal_needs_tracing");
  }
  if (asset.kind === "business_share") {
    addWarning(warnings, "business.charter_consent_required");
  }
  if (asset.kind === "crypto") {
    addWarning(warnings, "asset.crypto_evidence");
  }
}

export function claimPrice(caseDoc: MaritalPropertyCase, scenarioId: string, options?: ComputeOptions): number | null {
  return divide(caseDoc, scenarioId, options).claim_price;
}

export function warningsFor(caseDoc: MaritalPropertyCase, scenarioId: string, options?: ComputeOptions): string[] {
  return divide(caseDoc, scenarioId, options).warnings;
}

